internal/prehook/prehook.go

```diff
@@ -0,0 +1,175 @@
+// Package prehook implements a connection prehook mechanism, to handle any
+// pre-negotiation required by a remote node before the B2F protocol can
+// commence (e.g. packet node traversal).
+package prehook
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/la5nta/pat/internal/debug"
+	"github.com/la5nta/pat/internal/directories"
+	"golang.org/x/sync/errgroup"
+)
+
+var ErrConnNotWrapped = errors.New("connection not wrapped for prehook")
+
+type Script struct {
+	File string
+	Args []string
+	Env  []string
+}
+
+// Execute executes the prehook script on a wrapped connection.
+//
+// ErrConnNotWrapped is returned if conn is not wrapped.
+func (s Script) Execute(ctx context.Context, conn net.Conn) error {
+	if conn, ok := conn.(*Conn); ok {
+		return conn.Execute(ctx, s)
+	}
+	return ErrConnNotWrapped
+}
+
+type Conn struct {
+	net.Conn
+	br *bufio.Reader
+}
+
+// Verify returns nil if the given script file is found and valid.
+func Verify(file string) error { _, err := lookPath(file); return err }
+
+func lookPath(file string) (string, error) {
+	// Look in our custom location first
+	if p, err := exec.LookPath(filepath.Join(directories.ConfigDir(), "prehooks", file)); err == nil {
+		return p, nil
+	}
+	p, err := exec.LookPath(file)
+	if errors.Is(err, exec.ErrDot) {
+		return file, nil
+	}
+	return p, err
+}
+
+// Wrap returns a wrapped connection with the ability to execute a prehook.
+//
+// The returned Conn implements the net.Conn interface, and should be used in
+// place of the original throughout the lifetime of the connection once the
+// prehook script is executed.
+func Wrap(conn net.Conn) *Conn {
+	return &Conn{
+		Conn: conn,
+		br:   bufio.NewReader(conn),
+	}
+}
+
+func (p *Conn) Read(b []byte) (int, error) { return p.br.Read(b) }
+
+// Execute executes the prehook script, returning nil if the process
+// terminated successfully (exit code 0).
+func (p *Conn) Execute(ctx context.Context, script Script) error {
+	execPath, err := lookPath(script.File)
+	if err != nil {
+		return err
+	}
+	cmd := exec.CommandContext(ctx, execPath, script.Args...)
```

Comment:
How does the pre-hook script communicate its progress? Is exit code 0 the main/only way? I could see eventually expanding it to where the connection script can signal that the connection is ready, then Pat can signal when B2F is done and let the script tear down the connection. But I don't think that's necessary for v1.

Reply:
Yes, exit code 0 == success. Any other exit code is assumed to be a failure, and the connection is closed with an error presented to the user.

I see your point. We could maybe use a signal like `SIGCONT` for this, to keep stdio as transparent as possible. We would also need another mechanism to communicate that the connection is ready for B2F, as this is currently implemented by the child process exiting. But yes, maybe not necessary for v1.